File: 프로그래머스/3/49191. 순위/순위.cpp
#include <string>
#include <vector>
#define MAX 101

using namespace std;

vector<vector<bool>> dist(MAX, vector<bool>(MAX, false));

int solution(int n, vector<vector<int>> results) {
    int answer = 0;
    
    for(int i=0;i<results.size();i++){
        int win = results[i][0];
        int lose = results[i][1];
        dist[win][lose] = true;
    }
    
    for(int k=1;k<=n;k++){
        for(int i=1;i<=n;i++){
            for(int j=1;j<=n;j++){
                if(dist[i][k] && dist[k][j]){
                    dist[i][j] = true;
                }            
            }
        }
    }
    
    for(int i=1;i<=n;i++){
        int cnt=0;
        for(int j=1;j<=n;j++){
            if(dist[i][j] || dist[j][i])
                cnt++;
        }
        if(cnt == n-1)
            answer++;
    }
    
    return answer;
}

File: 프로그래머스/3/49191. 순위/순위.java
class Solution {
    public int solution(int n, int[][] results) {
        int answer = 0;
        
        boolean[][] graph = new boolean[n+1][n+1];
        for(int i=0;i<results.length;i++){
            int win = results[i][0];
            int lose = results[i][1];
            graph[win][lose] = true;
        }
        
        for(int k=1;k<=n;k++){
            for(int i=1;i<=n;i++){
                for(int j=1;j<=n;j++){
                    if(graph[i][k] && graph[k][j]){
                        graph[i][j] = true;
                    }
                }
            }
        }
        

        for(int i=1;i<=n;i++){
            int linkCount = 0;
            for(int j=1;j<=n;j++){
                if(graph[i][j] || graph[j][i]){
                    linkCount++;
                }
            }
            if(linkCount == n-1){
                answer++;
            }
        }
        
        return answer;
    }
}